### PR TITLE
runtime: stabilize worker_index

### LINKED ISSUE #8369: stabilize `worker_index`
I'd like to stabilize [`worker_index`](https://docs.rs/tokio/latest/tokio/runtime/fn.worker_index.html). No issues exist with the current API as far as I know and it seems fairly non-controversial

### PR DESCRIPTION
**Pull request by dajiaohuang**
## Motivation

	tokio::runtime::worker_index is currently gated on tokio_unstable, although #8369 requests stabilizing the API. Current-thread runtimes can have concurrent block_on callers, so only the thread that owns the driver has a worker context.

## Solution

Stabilize worker_index, document current-thread driver ownership, and add a regression test covering concurrent block_on calls. The existing behavior remains unchanged for multi-thread runtimes, blocking threads, and non-runtime callers.

Fixes: #8369

### CODE REVIEW

File: tokio/tests/rt_worker_index.rs
Comment (Darksonn) [inline tokio/tests/rt_worker_index.rs:1]: We're also missing a test that `Handle::block_on` always returns `None` even if it's the only call on a current-thread runtime, since this `block_on` method can never own the driver.

File: tokio/src/runtime/mod.rs [hunk 1]
```diff
@@ -563,43 +563,44 @@ cfg_rt! {
...
+    ///
+    /// Returns `None` when called from outside a runtime worker thread
+    /// (for example, from a blocking thread or a non-Tokio thread). On the
+    /// multi-thread runtime, the thread that calls [`Runtime::block_on`] is
+    /// not a worker thread, so this also returns `None` there.
+    ///
+    /// For the current-thread runtime and [`LocalRuntime`], this returns
+    /// `Some(0)` when called from the thread that currently owns the runtime
+    /// driver. If multiple threads call [`Runtime::block_on`] concurrently,
+    /// calls on threads that do not own the driver return `None`.
```
Comment (Darksonn) [inline tokio/src/runtime/mod.rs]: It's not possible for multiple threads to call `block_on` on a `LocalRuntime` since it's local. I think it would be nice to reword this slightly.

File: tokio/tests/rt_worker_index.rs
Comment (Darksonn) [inline tokio/tests/rt_worker_index.rs:1]: We're missing a test that `Runtime::block_on` always returns `None` on a multi-thread runtime, since this runtime always has the driver inside threads that it spawned itself.

Would be nice to also have a variation of this test that uses `#[tokio::test(flavor = "current_thread")]` and another using `#[tokio::test(flavor = "multi_thread")]`, just checking what happens if you invoke `worker_index` directly from a main function using the macro.

**Comment by Darksonn** [comment]
Thanks for the PR!

cc @rcoh

File: tokio/tests/rt_worker_index.rs
Comment (dajiaohuang) [reply tokio/tests/rt_worker_index.rs:1]: Added `worker_index_handle_block_on_current_thread`, which calls `rt.handle().block_on(...)` before any `Runtime::block_on` and asserts `None` on d76ee85.

Comment (dajiaohuang) [reply tokio/tests/rt_worker_index.rs:1]: Added `worker_index_tokio_test_current_thread` and `worker_index_tokio_test_multi_thread`; direct calls assert `Some(0)` and `None`. The focused test passes 9 stable tests and 10 with `RUSTFLAGS=--cfg tokio_unstable` on d76ee85.

File: tokio/src/runtime/mod.rs (hunk 1 above)
Comment (dajiaohuang) [reply tokio/src/runtime/mod.rs]: Reworded the documentation so concurrent driver ownership is scoped to the current-thread `Runtime`. It now states that `LocalRuntime` can only be driven from its owning thread, where calls inside `block_on` return `Some(0)`.

File: tokio/tests/rt_worker_index.rs
Comment (Darksonn) [reply tokio/tests/rt_worker_index.rs:1]: Can you also add non-macro version for multi-threaded runtime? Identical to `worker_index_current_thread` except for runtime flavor.

**Review by Darksonn — APPROVED** [review]
Thanks.
